### PR TITLE
[NTOS:MM] Stubplement cookie generation for drivers

### PR DESCRIPTION
**Pull request by learn-more**
Drivers compiled with a newer SDK check that their cookie is non-default.
This is not the same algorithm as Windows.

### CODE REVIEW

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 1]
```diff
@@ -2814,6 +2820,80 @@ MmCheckSystemImage(IN HANDLE ImageHandle,
...
+PVOID
+NTAPI
+LdrpFetchAddressOfSecurityCookie(PVOID BaseAddress, ULONG SizeOfImage)
+{
+    PIMAGE_LOAD_CONFIG_DIRECTORY ConfigDir;
+    ULONG DirSize;
+    PVOID Cookie = NULL;
+
+    /* Check NT header first */
+    if (!RtlImageNtHeader(BaseAddress)) return NULL;
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:2833]:
```suggestion
    if (!RtlImageNtHeader(BaseAddress))
        return NULL;
```

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 2]
```diff
@@ -2814,6 +2820,80 @@ MmCheckSystemImage(IN HANDLE ImageHandle,
...
+
+    /* Get the pointer to the config directory */
+    ConfigDir = RtlImageDirectoryEntryToData(BaseAddress,
+        TRUE,
+        IMAGE_DIRECTORY_ENTRY_LOAD_CONFIG,
+        &DirSize);
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:2839]:
```suggestion
    ConfigDir = RtlImageDirectoryEntryToData(BaseAddress,
                                             TRUE,
                                             IMAGE_DIRECTORY_ENTRY_LOAD_CONFIG,
                                             &DirSize);
```

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 3]
```diff
@@ -2814,6 +2820,80 @@ MmCheckSystemImage(IN HANDLE ImageHandle,
...
+
+    /* Check for sanity */
+    if (!ConfigDir ||
+        (DirSize != 64 && ConfigDir->Size != DirSize) ||
+        (ConfigDir->Size < 0x48))
+        return NULL;
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:2845]:
```suggestion
    if (!ConfigDir ||
        (DirSize != 64 && ConfigDir->Size != DirSize) ||
        (ConfigDir->Size < 0x48))
    {
        return NULL;
    }
```

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 4]
```diff
@@ -2814,6 +2820,80 @@ MmCheckSystemImage(IN HANDLE ImageHandle,
...
+NTAPI
+LdrpInitSecurityCookie(PLDR_DATA_TABLE_ENTRY LdrEntry)
+{
+    PULONG_PTR Cookie;
+    ULONG_PTR NewCookie;
+
+    /* Fetch address of the cookie */
+    Cookie = LdrpFetchAddressOfSecurityCookie(LdrEntry->DllBase, LdrEntry->SizeOfImage);
+
+    if (Cookie)
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:2871]:
```suggestion
    if (!Cookie)
        return NULL;
```
may be done similarly for the other if, and you gain one or two levels of indentation for the main code block below.

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 5]
```diff
@@ -3259,6 +3339,9 @@ MmLoadSystemImage(IN PUNICODE_STRING FileName,
     /* Write-protect the system image */
     MiWriteProtectSystemImage(LdrEntry->DllBase);
 
+    /* Initialize the security cookie (Win7 is not doing this yet!) */
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:3342]: You mean, it does it, but later?

File: ntoskrnl/mm/ARM3/sysldr.c (hunk 3 above)
Comment (HBelusca) [reply ntoskrnl/mm/ARM3/sysldr.c:2845]: Also, 0x48 == 72== `sizeof(IMAGE_LOAD_CONFIG_DIRECTORY)` for its 32-bit version (this is by the way platform specific, with `IMAGE_LOAD_CONFIG_DIRECTORY32` for 32-bits and `IMAGE_LOAD_CONFIG_DIRECTORY64` for 64 bits).

Comment (HBelusca) [reply ntoskrnl/mm/ARM3/sysldr.c:2845]: And 64 == `FIELD_OFFSET(IMAGE_LOAD_CONFIG_DIRECTORY, SecurityCookie)`

Comment (HBelusca) [reply ntoskrnl/mm/ARM3/sysldr.c:2845]: See also https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/apitests/loadconfig/common.c;hb=68182b2a52858eb0011df1467444e66e0b056f74#l47 that does a similar thing.
And these magic values should also be fixed in this function: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/ntdll/ldr/ldrinit.c;hb=68182b2a52858eb0011df1467444e66e0b056f74#l410

File: ntoskrnl/mm/ARM3/sysldr.c (hunk 5 above)
Comment (learn-more) [reply ntoskrnl/mm/ARM3/sysldr.c:3342]: Win10 is doing this, didnt check win8

**Comment by Extravert-ir** [comment]
Would be cool to enable security cookie in classpnp
https://github.com/reactos/reactos/blob/master/drivers/storage/class/classpnp/class.c#L231-L236

**Comment by learn-more** [comment]
> Nitpick for commit message module:
> 
> ```diff
> -[NTOS]
> +[NTOS:MM]
> ```

I think there is already enough nitpicking in this PR, guess it's better to make a huge pr because that will receive no comments.

File: ntoskrnl/mm/ARM3/sysldr.c (hunk 1 above)
Comment (learn-more) [reply ntoskrnl/mm/ARM3/sysldr.c:2833]: No, following the style of the surrounding code.

**Review by ColinFinck — APPROVED** [review]
LGTM. Just two remarks:
* This commit introduces architecture-specific `__rdtsc` in a non-architecture-specific file. The official alternative seems to be `NtGetTickCount`. However, the corresponding `KiGetTickCountHandler` trap handler currently always returns zero as the tick count, so this function isn't really an alternative right now.
* This commit also introduces `Ldrp` prefixes in an MM source file. I see that you copied most of the code from ntdll ldr, which is fine. Just not sure if we better rename the functions then to match the destination `Mi` prefix or keep things as is.

**Comment by HBelusca** [comment]
> LGTM. Just two remarks:
> 
>     * This commit introduces architecture-specific `__rdtsc` in a non-architecture-specific file. The official alternative seems to be `NtGetTickCount`. However, the corresponding `KiGetTickCountHandler` trap handler currently always returns zero as the tick count, so this function isn't really an alternative right now.
> 
>     * This commit also introduces `Ldrp` prefixes in an MM source file. I see that you copied most of the code from ntdll ldr, which is fine. Just not sure if we better rename the functions then to match the destination `Mi` prefix or keep things as is.

The corresponding user-mode Ldrp.... function does this:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/ntdll/ldr/ldrinit.c;hb=fd8080b0942e88502ca1735a0590f31e6d072e32#l464
i.e. calls NtQueryPerformanceCounter().
Since here we are in kernel-mode, and we can avoid probing etc.... we may as well call KeQueryPerformanceCounter() which does the correct job and abstract out the CPU-specific rdtsc .

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 6]
```diff
@@ -2814,6 +2820,83 @@ MmCheckSystemImage(IN HANDLE ImageHandle,
...
+    {
+        /* Check if it's a default one */
+        if ((*Cookie == DEFAULT_SECURITY_COOKIE) ||
+            (*Cookie == 0))
+        {
+            /* The address should be unique */
+            NewCookie = (ULONG_PTR)Cookie;
+
+            /* We just need a simple tick, don't care about precision and whatnot */
+            NewCookie ^= (ULONG_PTR)__rdtsc();
```
Comment (ColinFinck) [inline ntoskrnl/mm/ARM3/sysldr.c:2884]:
```suggestion
            NewCookie ^= (ULONG_PTR)KeQueryPerformanceCounter(NULL);
```

One-click fix based on @HBelusca's suggestion.

Comment (learn-more) [reply ntoskrnl/mm/ARM3/sysldr.c:2884]: I looked at that function, but wasnt sure about all the stuff it was doing...

**Comment by zefklop** [comment]
Can we move forward with this ? I see no reason to hold on.

**Comment by learn-more** [comment]
> Can we move forward with this ? I see no reason to hold on.

I did not re-test with the suggested change from colin.
If you want, feel free to apply that fix and re-test this.

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 7]
```diff
@@ -2814,6 +2820,80 @@ MmCheckSystemImage(IN HANDLE ImageHandle,
...
+
+            /* If the result is 0 or the same as we got, just add one to the default value */
+            if ((NewCookie == 0) || (NewCookie == *Cookie))
+            {
+                NewCookie = DEFAULT_SECURITY_COOKIE + 1;
```
[the remark this answers is not in the document]
Comment (ColinFinck) [reply ntoskrnl/mm/ARM3/sysldr.c:2889]: This situation is unlikely enough to not warrant a restart with `(ULONG_PTR)Cookie + 1` in a loop that could (theoretically) run infinitely.
Let's leave this part as is.

**Comment by ColinFinck** [comment]
> Can we move forward with this ? I see no reason to hold on.

Let's replace `__rdtsc` by `KeQueryPerformanceCounter` as suggested, try it, and if everything works, this PR should be good to go.

File: ntoskrnl/mm/ARM3/sysldr.c (hunk 6 above)
Comment (learn-more) [reply ntoskrnl/mm/ARM3/sysldr.c:2884]: `KeQueryPerformanceCounter` returns a `LARGE_INTEGER` so the one-click won't compile.

**Pull request merged**